### PR TITLE
Fjsp

### LINKED ISSUE #168: Add flexible shop scheduling environment
Add flexible shop scheduling environment.

### PR DESCRIPTION
**Pull request by LTluttmann**
## Description

Adding environment for the Flexible Job-Shop Scheduling Problem (FJSP) and the heterogeneous graph neural network (HGNN) described in Song et al. (https://ieeexplore.ieee.org/document/9826438) to solve it. Architecture was improved by adding residual connection and batch normalization. Average makespan on smallest instance type (10 jobs and 5 machines) is ~100 compared to ~106 reported in the paper.

## Motivation and Context

closes #168 

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: rl4co/envs/scheduling/fjsp/env.py [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+        # mask jobs that are done already
+        action_mask.add_(td["job_done"].unsqueeze(2))
+        # as well as jobs that are currently processed
+        action_mask.add_(td["job_in_process"].unsqueeze(2))
+
+        # mask machines that are currently busy
+        action_mask.add_(td["busy_until"].gt(td["time"].unsqueeze(1)).unsqueeze(1))
+
+        # exclude job-machine combinations, where the machine cannot process the next op of the job
+        next_ops_proc_times = (
```
Comment (fedebotu) [inline rl4co/envs/scheduling/fjsp/env.py:170]: Possible candidate for a `gather_by_index`? (same as function below)

File: rl4co/envs/scheduling/fjsp/env.py [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+        td["time"] = torch.where(step_complete, available_time, td["time"])
+
+        # this may only be set when the operation is finished, not when it is scheduled
+        # operation of job is finished, set next operation and flag job as being idle
+        curr_ops_end = td["finish_times"].gather(1, td["next_op"])
+        op_finished = torch.logical_and(
+            td["job_in_process"], curr_ops_end <= td["time"][:, None]
+        )
+        # check whether a job is finished, which is the case when the last operation of the job is finished
+        job_finished = torch.logical_and(op_finished, td["next_op"] == end_op_per_job)
```
Comment (fedebotu) [inline rl4co/envs/scheduling/fjsp/env.py:306]: [Minor] these `logical_and` might be replaced by `&` for readability?

File: rl4co/envs/scheduling/fjsp/generator.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+                "proc_times": proc_times,
+                "pad_mask": pad_mask,
+            },
+            batch_size=batch_size,
+        )
+
+        return td
+
+
+class FJSPFileGenerator(Generator):
```
Comment (fedebotu) [inline rl4co/envs/scheduling/fjsp/generator.py:150]: Great idea! We should do something similar in routing too

File: rl4co/envs/scheduling/fjsp/parser.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+def get_max_ops_from_files(files):
+    return max(map(get_n_ops_of_instance, files))
+
+
+def read(loc: Path, max_ops=None):
+    """
+    Reads an FJSPLIB instance.
+
+    Parameters
```
Comment (fedebotu) [inline rl4co/envs/scheduling/fjsp/parser.py:70]: [Minor] we are using the  following formatting for the docstrings

```python

"""
Args:
    loc: location of instance file
    
Returns:
    instance: the parsed instance
   """
```

File: rl4co/models/nn/ops.py [hunk 5]
```diff
@@ -35,3 +36,32 @@ def forward(self, x):
         else:
             assert self.normalizer is None, "Unknown normalizer type"
             return x
+
+
+class PositionalEncoding(nn.Module):
```
Comment (fedebotu) [inline rl4co/models/nn/ops.py:41]: Great, this was needed!

File: rl4co/models/zoo/hetgnn/model.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        policy_kwargs: Keyword arguments for policy
+        baseline_kwargs: Keyword arguments for baseline
+        **kwargs: Keyword arguments passed to the superclass
+    """
+
+    def __init__(
+        self,
+        env: RL4COEnvBase,
+        policy: HetGNNPolicy = None,
+        baseline: Union[REINFORCEBaseline, str] = "rollout",
```
Comment (fedebotu) [inline rl4co/models/zoo/hetgnn/model.py:27]: Does the original paper use "rollout" as baseline?

File: rl4co/tasks/eval.py [hunk 7]
```diff
@@ -169,23 +169,20 @@ def __init__(self, env, samples, softmax_temp=None, **kwargs):
         self.softmax_temp = softmax_temp
 
     def _inner(self, policy, td):
-        td = batchify(td, self.samples)
         out = policy(
             td.clone(),
             decode_type="sampling",
-            num_starts=0,
+            num_starts=self.samples,
+            multistart=True,
```
Comment (fedebotu) [inline rl4co/tasks/eval.py:176]: Won't this force the use of `select_start_nodes_fn`?

File: rl4co/utils/decoding.py [hunk 8]
```diff
@@ -360,6 +365,20 @@ def sampling(logprobs, mask=None):
 
         return selected
 
+    def _select_best(self, logprobs, actions, td: TensorDict, env: RL4COEnvBase):
+        def unbatchify_and_gather(x, idx):
```
Comment (fedebotu) [inline rl4co/utils/decoding.py:369]: Nice, worth another util? :rofl:

**Review by fedebotu — COMMENTED** [review]
Great quality code and lots of good changes too! 
I left some - mostly minor - comments here and there.

I don't have many comments on the scheduling, since I'm a newbie on that part. If @Junyoungpark you have some time, you may have a look too ;)

File: rl4co/models/zoo/hetgnn/model.py (hunk 6 above)
Comment (LTluttmann) [reply rl4co/models/zoo/hetgnn/model.py:27]: they actually use PPO, but using our RL4CO PPO implementation I could not get good results.

File: rl4co/utils/decoding.py (hunk 8 above)
Comment (LTluttmann) [reply rl4co/utils/decoding.py:369]: haha sure why not

File: rl4co/tasks/eval.py (hunk 7 above)
Comment (LTluttmann) [reply rl4co/tasks/eval.py:176]: great catch! Fix incoming

File: rl4co/models/zoo/hetgnn/model.py (hunk 6 above)
Comment (fedebotu) [reply rl4co/models/zoo/hetgnn/model.py:27]: Interesting; how about the critic neural network? Does it have the same structure as the HetGNN but with a value head?

**Review by fedebotu — APPROVED** [review]

**Review by fedebotu — APPROVED** [review]

File: rl4co/envs/scheduling/fjsp/generator.py (hunk 3 above)
Comment (cbhua) [reply rl4co/envs/scheduling/fjsp/generator.py:154]: Really nice idea! We were thinking of making this `generator.py` file a data manager, including generating data, loading from the file, and save to the file. This `FileGenerator()` is a nice implementation! Let's also apply this to other environments 😁

File: rl4co/envs/scheduling/fjsp/env.py [hunk 9]
```diff
@@ -0,0 +1,436 @@
...
+        self.num_mas = generator.num_mas
+        self.num_jobs = generator.num_jobs
+        self.n_ops_max = generator.max_ops_per_job * self.num_jobs
+        self.mask_no_ops = mask_no_ops
+        self._make_spec(self.generator)
+
+    def _set_seed(self, seed):
+        """Set the seed for the environment"""
+        rng = torch.manual_seed(seed)
+        self.rng = rng
```
Comment (cbhua) [inline rl4co/envs/scheduling/fjsp/env.py:91]: [Minor] this part is included in the `RL4COEnvBase()`: https://github.com/ai4co/rl4co/blob/684dbdcb6af83936a7733e814f1c9a42a280cab8/rl4co/envs/common/base.py#L267-L270

File: rl4co/envs/scheduling/fjsp/env.py [hunk 10]
```diff
@@ -0,0 +1,436 @@
...
+                # "op_spatial_enc": ops_spatial_enc,
+                "ops_sequence_order": ops_seq_order,
+            }
+        )
+
+        return td, n_ops_max
+
+    def _reset(self, td: TensorDict = None, batch_size=None) -> TensorDict:
+        device = td.device
+        self.to(device)
```
Comment (cbhua) [inline rl4co/envs/scheduling/fjsp/env.py:149]: [Minor] This part is included in the base class~ https://github.com/ai4co/rl4co/blob/684dbdcb6af83936a7733e814f1c9a42a280cab8/rl4co/envs/common/base.py#L135-L143

File: rl4co/envs/scheduling/fjsp/env.py [hunk 11]
```diff
@@ -0,0 +1,436 @@
...
+                "finish_times": finish_times,
+                "ma_assignment": ma_assignment,
+                "busy_until": busy_until,
+                "num_eligible": num_eligible,
+                "next_op": start_op_per_job.clone().to(torch.int64),
+                "ops_ma_adj": ops_ma_adj,
+                "op_scheduled": torch.full((*batch_size, n_ops_max), False),
+                "job_in_process": torch.full((*batch_size, self.num_jobs), False),
+                "reward": torch.zeros((*batch_size,), dtype=torch.float32),
+                "time": torch.zeros((*batch_size,)),
```
Comment (cbhua) [inline rl4co/envs/scheduling/fjsp/env.py:180]: Speaking about this initialize size for single dimension values: `[batch_size, 1]` or `[batch_size]` which one do you think it's better? 🤔

File: rl4co/envs/scheduling/fjsp/generator.py [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+        )
+
+        # TODO: This should be unnecessary due to multiplication with ma-ops edges
+        # proc_times[mask[:,None].expand_as(proc_times)] = 0
+        # remove proc_times for which there is no corresponding ma-ops connection
+        proc_times = proc_times * ma_ops_edges
+        return proc_times
+
+    def _generate(self, batch_size) -> TensorDict:
+        batch_size = [batch_size] if isinstance(batch_size, int) else batch_size
```
Comment (cbhua) [inline rl4co/envs/scheduling/fjsp/generator.py:103]: [Minor] This can also be removed since it is included in the base `Generator()` class

https://github.com/ai4co/rl4co/blob/684dbdcb6af83936a7733e814f1c9a42a280cab8/rl4co/envs/common/utils.py#L17-L19

File: rl4co/envs/scheduling/fjsp/generator.py [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+        pad_mask = pad_mask.ge(n_ops_batch[:, None].expand_as(pad_mask))
+
+        # determine the id of the end operation for each job
+        end_op_per_job = n_ope_per_job.cumsum(1) - 1
+
+        # determine the id of the starting operation for each job
+        # (bs, num_jobs)
+        start_op_per_job = torch.cat(
+            (torch.zeros((*batch_size, 1)), end_op_per_job[:, :-1] + 1), dim=1
+        )
```
Comment (cbhua) [inline rl4co/envs/scheduling/fjsp/generator.py:127]: [Minor] Convert to `int`.

File: rl4co/envs/scheduling/fjsp/env.py [hunk 14]
```diff
@@ -0,0 +1,436 @@
...
+            dtype=torch.int64,
+            low=-1,
+            high=self.n_ops_max,
+        )
+        self.reward_spec = UnboundedContinuousTensorSpec(shape=(1,))
+        self.done_spec = UnboundedDiscreteTensorSpec(shape=(1,), dtype=torch.bool)
+
+    @staticmethod
+    def render(td, idx):
+        return render(td, idx)
```
Comment (cbhua) [inline rl4co/envs/scheduling/fjsp/env.py:429]: I played with this render but I didn't get an output 🤔. Will this render plot actions? But this is in low priority, don't worry.

**Review by cbhua — APPROVED** [review]
Always great code! 👍 I tried to learn about the logic and pipeline of scheduling problem but I still couldn't go deep yet. But play with the model as a black-box works pretty good to me!

File: rl4co/models/zoo/hetgnn/model.py (hunk 6 above)
Comment (LTluttmann) [reply rl4co/models/zoo/hetgnn/model.py:27]: yes exactly!

File: rl4co/envs/scheduling/fjsp/env.py (hunk 14 above)
Comment (LTluttmann) [reply rl4co/envs/scheduling/fjsp/env.py:429]: hm that's weird, for me it works (in vscode). It should plot the (partial) schedule, like this for example:
image

File: rl4co/envs/scheduling/fjsp/env.py (hunk 11 above)
Comment (LTluttmann) [reply rl4co/envs/scheduling/fjsp/env.py:180]: puh that's a tough one :D I always decide case by case, trying to minimize the number of required reshaping operations

**Comment by LTluttmann** [comment]
thanks a ton for the valuable comments @fedebotu & @cbhua! I hope I addressed all issues. Merging now
